scipy/optimize/tests/test_linprog.py

```diff
@@ -442,14 +442,13 @@ def f(c, A_ub=None, b_ub=None, A_eq=None, b_eq=None, bounds=None):
         bounds = np.column_stack((lb, ub))
         b_eq = A_eq * x_valid
 
+        # simplex methods now convert sparse to dense array and warn gh-14448
         if self.method in {'simplex', 'revised simplex'}:
-            # simplex and revised simplex should raise error
-            with assert_raises(ValueError, match=f"Method '{self.method}' "
-                               "does not support sparse constraint matrices."):
+            with assert_warns(scipy.sparse.SparseEfficiencyWarning):
                 linprog(c=c, A_eq=A_eq, b_eq=b_eq, bounds=bounds,
```

Comment:
To address the [comment above](https://github.com/scipy/scipy/pull/14488/files#r680087688) about unit testing, you can just check that `linprog` correctly solves this problem.

Reply:
Sorry, I'm not very familiar with writing tests...would this be the way to do it? Thanks for your patience

        if self.method in {'simplex', 'revised simplex'}:
            #keep assert_warns check or no? 
            with assert_warns(scipy.sparse.SparseEfficiencyWarning):
                linprog(c=c, A_eq=A_eq, b_eq=b_eq, bounds=bounds,
                        method=self.method, options=self.options)

            #this is the correct way to check the problem is solved correctly? 
            res = linprog(c=c, A_eq=A_eq, b_eq=b_eq, bounds=bounds,
                          method=self.method, options=options)
            assert res.success

        else:
            # other methods do not warn
            options = {**self.options}
            if self.method in {'interior-point'}:
                options['sparse'] = True

            res = linprog(c=c, A_eq=A_eq, b_eq=b_eq, bounds=bounds,
                          method=self.method, options=options)
            assert res.success

Reply:
> #keep assert_warns check or no? 

Yes, you want to check for the warning. You can condense things a little:

```
    if self.method in {'simplex', 'revised simplex'}:
        # Old simplex methods convert sparse to dense
        with assert_warns(scipy.sparse.SparseEfficiencyWarning):
            res = linprog(c=c, A_eq=A_eq, b_eq=b_eq, bounds=bounds,
                          method=self.method, options=self.options)
    else:
        options = {**self.options}
        if self.method in {'interior-point'}:
            options['sparse'] = True

        res = linprog(c=c, A_eq=A_eq, b_eq=b_eq, bounds=bounds,
                      method=self.method, options=options)

        assert res.success
```

> #this is the correct way to check the problem is solved correctly? 

I suggested that you add a check for a correct solution because there is a typo above `A_ub = A_eq.toarray()` that this test should catch (by failing). Does this test fail currently and pass when you fix the typo? If not (e.g. it passes with the typo) then it certainly needs to be strengthened. Even if it _does_ (fail now and pass when you fix the typo), it probably should be strengthened because `res.success` does not ensure that that `linprog` solved the problem that you intended it to solve.

I'd suggest that the test check whether the solution `res.x` is the _correct solution to the original problem_. You can do that by changing the problem to one for which you already know the answer (e.g. it is published somewhere or you have some other an external reference) or you solve the problem with one of the solvers that you are not changing the behavior of (e.g. one of the HiGHS solvers. Since you are not changing their behavior, you can use them to test the behavior of the solvers you _are_ changing - the old simplex solvers.)